capsules/src/bmp280.rs

```diff
@@ -0,0 +1,469 @@
+            State::InitId
+            | State::InitResetting
+            | State::InitWaitingReady
+            | State::InitReadingCalibration => Err(ErrorCode::BUSY),
+            State::Idle(calibration) => {
+                self.buffer.take().map_or(Err(ErrorCode::NOMEM), |buffer| {
+                    // todo: use bitfield crate
+                    // forced mode, oversampling 1
+                    let val = 0b00100001;
+                    let (ret, new_state) = match self.i2c.write(buffer, Register::CTRL_MEAS, [val])
+                    {
+                        Ok(()) => (Ok(()), State::Configuring(calibration)),
+                        Err((_e, buffer)) => {
+                            self.i2c.disable();
+                            self.buffer.replace(buffer);
+                            (Err(ErrorCode::FAIL), State::Idle(calibration))
+                        }
+                    };
+                    self.state.set(new_state);
+                    ret
+                })
+            }
+            State::Configuring(_)
+            | State::WaitingForAlarm(_)
+            | State::Waiting(_)
+            | State::Reading(_) => Err(ErrorCode::BUSY),
+            State::Error => Err(ErrorCode::FAIL),
+            State::IrrecoverableError => Err(ErrorCode::NODEVICE),
+        }
+    }
+
+    fn handle_alarm(&self) {
+        match self.state.get() {
+            State::WaitingForAlarm(calibration) => self.buffer.take().map_or_else(
+                || debug!("BMP280 No buffer available!"),
+                |buffer| {
+                    let new_state = match self.check_ready(buffer) {
+                        Ok(()) => State::Waiting(calibration),
+                        Err((_e, buffer)) => {
+                            self.i2c.disable();
+                            self.buffer.replace(buffer);
+                            State::Idle(calibration)
+                        }
+                    };
+                    self.state.set(new_state)
+                },
+            ),
+            other => {
+                debug!("BMP280 received unexpected alarm in state {:?}", other);
+            }
+        }
+    }
+
+    fn arm_alarm(&self) {
+        // Datasheet says temp oversampling=1 makes a reading typically take 5.5ms.
+        // (Maximally 6.4ms).
+        let delay = self.alarm.ticks_from_us(6400);
+        self.alarm.set_alarm(self.alarm.now(), delay);
+    }
+
+    fn check_ready(
+        &self,
+        buffer: &'static mut [u8],
+    ) -> Result<(), (i2c::Error, &'static mut [u8])> {
+        self.i2c.read(buffer, Register::STATUS, 1)
+    }
+}
+
+enum I2cOperation {
+    Read {
+        addr: Register,
+        count: u8,
+        fail_state: State,
+    },
+    Write {
+        addr: Register,
+        data: u8,
+        fail_state: State,
+    },
+    Disable,
+}
+
+impl I2cOperation {
+    fn read(addr: Register, count: u8, fail_state: State) -> Self {
```

Comment:
Given `I2cOperation` is module-private by default, and these methods don't do anything other than construct the enum variants with the passed arguments, you can just construct the variants below (you already do that for `I2cOperation::disable` anyways.

Reply:
I added those as a syntax sugar. Without those,
```
I2cOperation::read(Register::TEMP_MSB, 3, State::Idle(calibration));
```

becomes
```
I2cOperation::Read(addr: Register::TEMP_MSB, count: 3, state: State::Idle(calibration));
```

I don't have a better reason to keep those, other than that it will be annoying to write the pressure measurement code without them.